### ISSUE TITLE
Move to Mio/Tokio

### ISSUE THREAD

**Issue by Noah-Kennedy** [cloudflare/boringtun#217]
There is no reason we should be rolling our own epoll and kqueue. Let's just migrate to tokio or mio.

**Comment by Noah-Kennedy** [comment]
I have some concerns about the amount of `unsafe` in here. Much of this is custom crap that can be ripped out and replaced easily.

**Comment by zaid3tech** [comment]
Should we expect these changes in near future?

**Comment by Noah-Kennedy** [comment]
@zaid3tech yes, I don't have a set date to give you, but you should expect this soon.

**Comment by zaid3tech** [comment]
Thanks : )

**Comment by ngc0202** [comment]
Is this still in the works? I'd love to see it implemented.

**Comment by Noah-Kennedy** [comment]
Yes, I just have had reduced bandwidth due to some other things.

**Comment by ExternalReality** [comment]
Hi all,

I hope all is well. Question, is the plan to just to replace the custom events system with mio, or to use Tokio runtime task scheduling etc?

**Comment by Noah-Kennedy** [comment]
The plan is to use tokio.

**Comment by ExternalReality** [comment]
Hi @Noah-Kennedy, I know that this work has been on the backlog for sometime now. Any information on where this stands. I ask because any contribution my be dependent on something that might significantly change based on this rework. :-)

**Comment by lz1998** [comment]
Hi @Noah-Kennedy , I am trying to move to `tokio`. Can you create a `tokio` branch which I can submit PR to?

https://github.com/lz1998/wg-rs

**Comment by lodrem** [comment]
I just wanted to check in on when the migration will be completed. FYI, I've reimplemented it using Tokio and hope that it could help with the migration: https://github.com/lodrem/wiretun
